### PR TITLE
rollout: Report health for all cases

### PR DESCRIPTION
**Pull request by gvso**
This reports the health for all possible cases. In particular, it reports inconclusive diagnosis (no enough requests) and healthy diagnosis without enough elapsed time.

### CODE REVIEW

File: internal/health/health.go [hunk 1]
```diff
@@ -82,16 +82,20 @@ func Diagnose(ctx context.Context, healthCriteria []config.HealthCriterion, actu
...
 
 		isMet := isCriteriaMet(criteria.Metric, criteria.Threshold, value)
+		result := CheckResult{Threshold: criteria.Threshold, ActualValue: value}
 
 		// For unmet request count, return inconclusive and empty results.
 		if !isMet && criteria.Metric == config.RequestCountMetricsCheck {
 			logger.Debug("unmet criterion")
 			diagnosis = Inconclusive
-			results = nil
+			// TODO: This will be an issue if the request count is not the first
```
Comment (gvso) [inline internal/health/health.go:91]: Potentially solvable easily if we store Metric as field in `CheckResult`

File: internal/health/health.go [hunk 2]
```diff
@@ -82,16 +82,20 @@ func Diagnose(ctx context.Context, healthCriteria []config.HealthCriterion, actu
...
+			// criterion in the array since the caller can assume the order
+			// of the check results is the same as the order of the health
+			// criteria.
+			results = []CheckResult{result}
```
Comment (gvso) [inline internal/health/health.go:95]: Potentially solvable easily if we store Metric as field in CheckResult

**Comment by gvso** [comment]
The two new reports look like:

```yaml
rollout.cloud.run/lastHealthReport: |-
      status: healthy, but no enough time since last rollout
      metrics:
      - error-rate-percent: 0.00 (needs 1.00)
      - request-count: 5 (needs 0)
      lastUpdate: 2020-08-18T15:10:16-04:00
rollout.cloud.run/lastRollout: '2020-08-18T15:09:35-04:00'
```

```yaml
rollout.cloud.run/lastHealthReport: |-
      status: inconclusive
      metrics:
      - request-count: 0 (needs 100)
      lastUpdate: 2020-08-18T15:35:00-04:00
```

**Comment by ahmetb** [comment]
🤔 it's sort of nice to omit other metrics in the status report when they're not relevant, but I suspect that might cause more confusion than it's helping. (e.g. people might think the metrics they configured aren't being taken into consideration by the program).

**Comment by gvso** [comment]
I had a dilemma between your point and the user not knowing what `inconclusive` means when the other criteria seems fine. Maybe we should do `inconclusive, no enough requests`?

File: internal/health/health.go [hunk 3]
```diff
@@ -82,16 +82,20 @@ func Diagnose(ctx context.Context, healthCriteria []config.HealthCriterion, actu
 		}
 
 		isMet := isCriteriaMet(criteria.Metric, criteria.Threshold, value)
+		result := CheckResult{Threshold: criteria.Threshold, ActualValue: value}
 
 		// For unmet request count, return inconclusive and empty results.
 		if !isMet && criteria.Metric == config.RequestCountMetricsCheck {
 			logger.Debug("unmet criterion")
```
Comment (ahmetb) [inline internal/health/health.go:89]: duplicate log line

File: internal/health/health.go [hunk 4]
```diff
@@ -82,16 +82,20 @@ func Diagnose(ctx context.Context, healthCriteria []config.HealthCriterion, actu
...
 			diagnosis = Inconclusive
-			results = nil
+			// TODO: This will be an issue if the request count is not the first
+			// criterion in the array since the caller can assume the order
+			// of the check results is the same as the order of the health
+			// criteria.
+			results = []CheckResult{result}
 			break
```
Comment (ahmetb) [inline internal/health/health.go:96]: return all CheckResults here.
if there's unhealthy despite low req count, that's still unhealthy (+add a test case for this)

File: internal/rollout/rollout_test.go [hunk 5]
```diff
@@ -293,7 +305,15 @@ func TestUpdateService(t *testing.T) {
 			lastReady: "test-002",
 			healthCriteria: []config.HealthCriterion{
 				{Metric: config.RequestCountMetricsCheck, Threshold: 1500},
-				{Metric: config.ErrorRateMetricsCheck, Threshold: 0.95},
+				{Metric: config.ErrorRateMetricsCheck, Threshold: 5},
+			},
+			outAnnotations: map[string]string{
+				rollout.StableRevisionAnnotation:    "test-001",
+				rollout.CandidateRevisionAnnotation: "test-002",
+				rollout.LastHealthReportAnnotation: "status: inconclusive\n" +
```
Comment (ahmetb) [inline internal/rollout/rollout_test.go:313]: potentially it's gonna be a hassle to provide full values here.

File: internal/health/health.go (hunk 4 above)
Comment (gvso) [reply internal/health/health.go]: I gave `Unhealthy` precedence over `Inconclusive` and appended all results

File: internal/rollout/rollout_test.go (hunk 5 above)
Comment (gvso) [reply internal/rollout/rollout_test.go:313]: Yeah, let's work on this in another PR

**Pull request merged**